### ISSUE TITLE
[data] Add backpressure from clients to streaming_split coordinator

### ISSUE THREAD

**Issue by stephanie-wang** [ray-project/ray#39595]
### What happened + What you expected to happen

streaming_split is used for training use cases, where the output should get split among N trainers. Each trainer chooses when to request more blocks from the streaming_split coordinator's buffer. Once a block reaches the coordinator's buffer or the trainer's buffer, we no longer track the memory store usage. Also, the coordinator only synchronizes the trainers at each epoch.

This leads to two problems:
- higher object store memory usage than expected (because some blocks are missing from the streaming executor's backpressure accounting)
- a slower trainer can fall behind and, with locality scheduling enabled, cause blocks to build up on its node

We should add backpressure from the trainers to the coordinator to avoid this situation. This will effectively guarantee that all trainers stay within X rows of each other.

### Versions / Dependencies

3.0dev

### Reproduction script

-

### Issue Severity

None

**Comment by stephanie-wang** [comment]
In practice, this may not be a concern for training use cases, since data-parallel workers will often synchronize on every batch anyway.

**Comment by ericl** [comment]
Actually @stephanie-wang I don't think this is the same issue as #39600 --- for the imbalance issue reported there, the imbalance is in `OpState.output_queue`, not the data prefetched to the consumer.

**Comment by stephanie-wang** [comment]
Ah I see.

Actually, I'll change the priority of this one to p1 since we can assume trainers are usually synchronized, but we should really fix the memory accounting issue/include consumers in backpressure calculation separately.

**Comment by ericl** [comment]
Yup I think that makes sense. Btw, for the tracking I think we should also make sure to include the memory usage for the consumer/output in the progress bar, otherwise it's pretty hard to see where the memory is going.

**Comment by anyscalesam** [comment]
tracked in data-tracker
